### DISCUSSION TITLE
Trouble With Sound Stuttering ON XIAO ESP32S3 from Streaming MP3

### DISCUSSION THREAD

**Discussion by jimruxton** [pschatzmann/arduino-audio-tools#1099] [Q&A] [answered]
I'm testing out the example streams-url_mp3_helix-i2s. I feel like I am setting the wrong compiler flags or setting the wrong values in AudioConfig.h . In the program streams-url_mp3_helix-i2s I am reading an mp3 file that is streaming over Ethernet to my router. The sound is very choppy. I tried increasing the I2S_BUFFER_COUNT and I2S_BUFFER_SIZE but that didn't change anything. I also tried making the values of URL_STREAM_BUFFER_COUNT and STACK_SIZE larger but again no change. I also tried lowering them but it sounded the same as if the values weren't affecting anything. Could they be getting set somewhere else?  I don't think my problem is signal strength as if I put the esp32 near the router I get an RSSI of -11 and still get choppy audio.  I know it isn't the stream coming out of my router since I can go far with my moblie phone and play the stream without glitches

 with an RSSI of -55 . Does anyone have any thoughts what the problem could be either with the way I am compiling or WiFi configuration in Audio Tools? Attached is a screenshot of my compiler flags.
![Screenshot from 2023-12-12 12-19-23](https://github.com/pschatzmann/arduino-audio-tools/assets/29495434/b1266741-73ff-4274-90c8-092c95cb0e8f)

**Comment by pschatzmann** [comment, marked as answer]
I don't know where you are located: Did you try with a local url.
Your test with the mobile phone does not mean anything: on the mobile phone megabytes are buffered before the audio starts to play.
On microconrollers it is vital that the data is provided at a constant (fast enough speed) and any short drop will be audible.
You should also [measure the network speed with the audiotools](https://www.pschatzmann.ch/home/2023/02/28/arduino-audiotools-esp32-vs-rp2040-pico-w/).

So my gut feeling is that you have an infrastructure problem. You can double check e.g. if you just serve an mp3 file from a local webserver where the local network traffic is under your control.

ps. I assume that you were not so crasy to replace the demo url with one that uses some extreme or insane bitrates

**Comment by jimruxton** [comment]
Thanks. I will try doing the measurements but I was already serving an mp3
file from a local webserver. On my laptop I ran :

> sudo python3 -m http.server 80  (which was connected by Ethernet to my
> router)
>
And on the esp32 I ran the following code. (Alicia.mp3 was in the directory
where the http.server was running)

/**
 * @file streams-url_mp3-i2s.ino
 * @author Phil Schatzmann
 * @brief decode MP3 stream from url and output it on I2S
 * @version 0.1
 * @date 2021-96-25
 *
 * @copyright Copyright (c) 2021
 */

// install https://github.com/pschatzmann/arduino-libhelix.git

#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"


URLStream url("SSID","password");
I2SStream i2s; // final output of decoded stream
EncodedAudioStream dec(&i2s, new MP3DecoderHelix()); // Decoding stream
StreamCopy copier(dec, url); // copy url to decoder


void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup i2s
  auto config = i2s.defaultConfig(TX_MODE);
  // you could define e.g your pins and change other settings
  //config.pin_ws = 10;
  //config.pin_bck = 11;
  //config.pin_data = 12;
  //config.mode = I2S_STD_FORMAT;
  i2s.begin(config);

  // setup I2S based on sampling rate provided by decoder
  dec.setNotifyAudioChange(i2s);
  dec.begin();

// mp3 radio
  url.begin("http://192.168.1.5/Alicia.mp3","audio/mp3");

}

void loop(){
  copier.copy();
}

**Comment by jimruxton** [comment]
Just a quick note to say I also tried running the AudioLogger with "warning"

**Comment by jimruxton** [comment]
I ran the program to measure the bitrate streams-url_mp3-measuring.ino . I got the following bitrates. It looks plenty fast to me. The test was done using the same mp3 file and network infrastructure as described above.

```
18:40:12.352 -> ==> Bytes per second: 199000
18:40:13.481 -> ==> Bytes per second: 211000
18:40:15.009 -> ==> Bytes per second: 154000
18:40:16.204 -> ==> Bytes per second: 193000
18:40:17.432 -> ==> Bytes per second: 191000
18:40:18.561 -> ==> Bytes per second: 209000
18:40:20.188 -> ==> Bytes per second: 145000
18:40:21.549 -> ==> Bytes per second: 172000
18:40:22.944 -> ==> Bytes per second: 167000
18:40:24.140 -> ==> Bytes per second: 192000
18:40:25.369 -> ==> Bytes per second: 190000
18:40:26.896 -> ==> Bytes per second: 150000
18:40:28.062 -> ==> Bytes per second: 204000
18:40:29.192 -> ==> Bytes per second: 203000
18:40:30.655 -> ==> Bytes per second: 162000
18:40:31.883 -> ==> Bytes per second: 188000
18:40:33.012 -> ==> Bytes per second: 210000
18:40:34.107 -> ==> Bytes per second: 213000
18:40:35.268 -> ==> Bytes per second: 204000
18:40:36.861 -> ==> Bytes per second: 146000
18:40:38.123 -> ==> Bytes per second: 189000
18:40:39.584 -> ==> Bytes per second: 159000
18:40:40.780 -> ==> Bytes per second: 196000
18:40:41.909 -> ==> Bytes per second: 211000
18:40:43.004 -> ==> Bytes per second: 209000
18:40:44.265 -> ==> Bytes per second: 187000
18:40:45.692 -> ==> Bytes per second: 161000
18:40:46.953 -> ==> Bytes per second: 187000
18:40:48.247 -> ==> Bytes per second: 181000
18:40:49.441 -> ==> Bytes per second: 200000
18:40:50.736 -> ==> Bytes per second: 179000
18:40:51.931 -> ==> Bytes per second: 198000
18:40:53.092 -> ==> Bytes per second: 202000
18:40:54.320 -> ==> Bytes per second: 192000
18:40:55.681 -> ==> Bytes per second: 172000
18:40:56.876 -> ==> Bytes per second: 195000
18:40:58.137 -> ==> Bytes per second: 185000
18:40:59.330 -> ==> Bytes per second: 195000
```

**Comment by pschatzmann** [comment]
I am not quite sure: In order to never stutter all values might need to be above 44200* 2 channels * 2 bytes (=16 bits)) = 176400 bytes / sec
Did you increase the I2S buffer to at least 1024 * 200 which would be a bit more then 1 second ?

What is the bitrate of your mp3 ? Did you try to convert it to standard e.g. with Audacity ?

**Comment by jimruxton** [comment]
Thanks again Phil for your suggestions. This is what ffprobe gives me on the mp3 file 
```
Duration: 00:03:29.50, start: 0.000000, bitrate: 320 kb/s
  Stream #0:0: Audio: mp3, 44100 Hz, stereo, fltp, 320 kb/s 
```
So a couple things , it is a high bitrate but I thought if it was less than the transmission rates I am getting , lowest was 150kbytes/sec or 1200kb/s  . Meanwhile I will try converting the file to a lower bit rate. Also I see the format is fltp , although in between stutters when the sound plays it sounds ok so I guess the decoder can handle this format ok.

When I tried to raise the i2s buffer size to 200 I got a message in the monitor saying it had to be between 128 and 2 . However anything above 50 and I get nothing in fact the espe32 keeps rebooting with buffer sizes above 64 . Below is the monitor output with a buffer size of 70. It is continuously rebooting. I realize this probably isn't audio tools related but I thought it may provide other clues. 

```
ELF file SHA256: 48d4502fc0d3990c

Rebooting...
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0xc (RTC_SW_CPU_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x42093a36
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x3ac
load:0x403c9700,len:0x9a8
load:0x403cc700,len:0x2920
entry 0x403c98b8
[I] I2SConfig.h : 104 - rx/tx mode: TX_MODE
[I] I2SConfig.h : 105 - port_no: 0
[I] I2SConfig.h : 106 - is_master: Master
[I] I2SConfig.h : 107 - sample rate: 44100
[I] I2SConfig.h : 108 - bits per sample: 16
[I] I2SConfig.h : 109 - number of channels: 2
[I] I2SConfig.h : 110 - i2s_format: I2S_STD_FORMAT
[I] I2SConfig.h : 112 - auto_clear:1
[I] I2SConfig.h : 119 - buffer_count:70
[I] I2SConfig.h : 120 - buffer_size:1024
[I] I2SConfig.h : 124 - pin_bck: 2
[I] I2SConfig.h : 126 - pin_ws: 4
[I] I2SConfig.h : 128 - pin_data: 3
[I] AudioEncoded.h : 249 - virtual void audio_tools::EncodedAudioPrint::setNotifyAudioChange(audio_tools::AudioBaseInfoDependent&)

abort() was called at PC 0x42081f7f on core 1


Backtrace: 0x4037743e:0x3fcebd00 0x4037cef9:0x3fcebd20 0x40382c8d:0x3fcebd40 0x42081f7f:0x3fcebdc0 0x42081fc6:0x3fcebde0 0x42082307:0x3fcebe00 0x42082266:0x3fcebe20 0x42082361:0x3fcebe40 0x42003abf:0x3fcebe60 0x420048e0:0x3fcebe80 0x42005472:0x3fcebea0 0x42008472:0x3fcebec0 0x420107f2:0x3fcebf70
 
```

**Comment by pschatzmann** [comment]
Using Audacity terms: a standard bitrate (170-210 kbps) should work w/o any changes.
If you want to use a higher bitrate (extreme or insane) you will need to increase the memory settings as described in the Readme of the helix project.

**Comment by jimruxton** [comment]
Thanks Phil. I just tested with a 64 kbps mp3 file and it worked fine. Would you have any thoughts as to why when increasing the i2s buffer_count above 64 the program no longer works. Would it help to implement PSRAM in some way at that point?

**Comment by pschatzmann** [comment]
My guess would be that you just run out of RAM, so [PSRAM](https://www.pschatzmann.ch/home/2022/05/30/esp32-and-psram/) might help, but I never checked if this is working.
